### ISSUE TITLE
Which ports does lego actually need/use?

### ISSUE THREAD

**Issue by shurcooL** [xenolf/lego#47]
This is another question/clarification.

From reading the entire README, my understanding is that `lego` only needs to use port 443 by default, unless you use a custom port, then you need to forward 443 to that.

From https://www.guyrutenberg.com/2015/11/07/getting-started-with-lets-encrypt-tutorial/:

> The next step is to issue the certificate and prove to Let’s Encrypt that you have some control over the domain. The client supports two methods to perform the validation. The first one is the standalone server. It works by setting up a webserver on port 443, and responding to a challenge from the Let’s Encrypt servers. However, if you already have your own web-server running on port 443 (the default for TLS/SSL), you would have to temporarily shut it down.

Which seems to confirm that idea.

However, commit https://github.com/xenolf/lego/commit/c4add3c81ed1bac4f8df20d9ca76e47521a82401 says:

> Assumes the `lego` binary has permission to bind to ports 80 and 443.

This is the _only_ mention of port 80 that I see in the README.

Is that accurate? Does it actually need both 80 and 443 (meaning I need to temporarily stop whatever's running on both 80 and 443 when running `lego`)? Or does it only need 443, and mentioning port 80 is unintended? /cc @mholt

Whatever the answer is, I think it would be helpful to cover in a little more detail that given ports need to be available temporarily while `lego` is running, and that you need to shutdown something else that might be running there. Just to let people know that it's normal procedure (and maybe suggest alternatives, etc.).

Thanks again, this project is great!

**Comment by shurcooL** [comment]
Well, from trying it, it looks like it *does* use/need port 80 in addition to 443:

```
2015/12/08 01:47:41 [INFO] acme: Trying to solve HTTP-01
2015/12/08 01:47:41 [www.example.com] Could not obtain certificates
	Could not start HTTP server for challenge -> listen tcp :80: bind: permission denied
```

I find that surprising because of the following parts of the README:

> - Pass the --port option and specify a custom port to bind to. In this case you have to forward port 443 to this custom port.

Mentions only 443, not 80. It's not even --https-port, just --port, which makes you think there's only ever going to be one.

> ```
>   --port                               Challenges will use this port to listen on. Please make sure to forward port 443 to this port on your machine. Otherwise use setcap on the binary
> ```

Same.

> ```Go
> 
> // A client facilitates communication with the CA server. This CA URL is
> // configured for a local dev instance of Boulder running in Docker in a VM.
> // We specify an optPort of 5001 because we aren't running as root and can't
> // bind a listener to port 443 (used later when we attempt to pass challenge).
> client, err := acme.NewClient("http://192.168.99.100:4000", &myUser, rsaKeySize, "5001")
> if err != nil {
>   log.Fatal(err)
> }
> ```

Once again, only 443 is mentioned, and it's made to sound like the _only_ port.

**Comment by xenolf** [comment]
The README has not been updated to reflect the latest changes. Thanks for reminding me to do that.
Lego uses the ports 80 and 443 depending on which challenge is being solved. HTTP-01 needs port 80 and TLS-SNI-01 needs port 443 as specified in the ACME spec. 

The --port option was included to allow people to run lego without root and/or when there already is a webserver running which is occupying the mentioned ports. In case you pass the --port option to lego it changes the listening port of any challenge which needs a listening port to the port you specified.

**Comment by shurcooL** [comment]
> The README has not been updated to reflect the latest changes. Thanks for reminding me to do that.

Thanks!

> Lego uses the ports 80 and 443 depending on which challenge is being solved. HTTP-01 needs port 80 and TLS-SNI-01 needs port 443 as specified in the ACME spec.
> 
> In case you pass the --port option to lego it changes the listening port of any challenge which needs a listening port to the port you specified.

Oh, really. So if I set `--port=5000`, does that mean both both HTTP-01 and TLS-SNI-01 will get answered on port 5000? Do you still need to forward 433 to 5000, or is that part outdated?

**Comment by xenolf** [comment]
Lego gets a list of challenges from the server. We choose the first challenge in this list which we can solve. Depending on which challenge that is (the list is always randomized - at the moment 50/50 for HTTP-01 and TLS-SNI-01) we then start a server on either port 80 or 443 if no `--port` option was spcified.
If you pass `--port=5000` we will use that port instead and depending on the challenge either serve TLS-SNI-01 or HTTP-01 on that port. After the challenge is solved we tear the server down and go on. 

In case of HTTP-01 you need to forward connections that have a request path starting with `/.well-known/acme/` from port 80 to port 5000. 
For TLS-SNI-01 you need to forward port 443 to port 5000 as we need to be able to respond to TLS handshakes - there is no HTTP involved here.
Also note that TLS-SNI-01 will not work when SSL termination is happening somewhere else.

**Comment by shurcooL** [comment]
That's very clear and I understand now, thanks.

**Comment by xenolf** [comment]
Fixed in 6314bea
